### ISSUE TITLE
Input ops fed networks operate considerably slower than direct feed ones

### ISSUE THREAD

**Issue by danielschonfeld** [tensorflow/tensorflow#4732]
### What related GitHub issues or StackOverflow threads have you found by searching the web for your problem?
http://stackoverflow.com/questions/39794149/when-introducing-the-optimizer-variables-under-variable-scope-get-recreated-twic

### Environment info
Tried both OS X and Linux (Ubuntu 16)
On both utilizing CPU only.

### If possible, provide a minimal reproducible example (We usually don't have time to read hundreds of lines of your code)
Any simple operation (i.e. calculating logits with a 3 layer deep network with simple regression in each layer) in which the data was fed with either a `parse_example`, `parse_single_example`, building a `CustomRunner` that feeds a `RandomShuffleQueue` or utilizing `QueueRunner`.  In all of those the operation and time to complete 1 epoch took considerably more than if I were to save the data in a bumpy array and feed it during a call to `sess.run()` with the `feed_dict`.

### What other attempted solutions have you tried?
Tried to score the web for solutions and ran into a couple blog posts describing the problem but can't seem to find solutions.

### Logs or other output that would be helpful
Tried this on both TF 0.10 and the latest RC for 0.11.

**Comment by tatatodd** [comment]
@ebrevdo might have some thoughts on this.

**Comment by yaroslavvb** [comment]
I looked at some timelines in https://github.com/tensorflow/tensorflow/issues/4740 and [here](http://stackoverflow.com/questions/39840323/benchmark-of-howto-reading-data).

On official MNIST examples QueueDequeueMany is the bottleneck, whether it's reading using `parse_single_example` or using `slice_input_producer`, with or without the GPU, taking 10x longer than entire duration of step when using `feed_dict` with same data size

- examples/how_tos/reading_data/fully_connected_reader.py
- examples/how_tos/reading_data/fully_connected_preloaded.py
- examples/how_tos/reading_data/fully_connected_preloaded_var.py

**Comment by tatatodd** [comment]
(see #4740 for an attached timeline)

**Comment by ebrevdo** [comment]
`QueueDequeueMany` takes a long time when it's waiting for enough stuff to get put on the queue.  It's not computation cycles that are slowing it down.

You need multiple reader threads all feeding the queue in order to unblock it.  If you're using `tf.train.batch` or its friends, in TensorBoard you'll see that the queue is never full (and it should be).  Increase the number of threads you use; this is a simple argument in `tf.train.batch` and friends.

**Comment by yaroslavvb** [comment]
Ah, good point, increasing capacity, number of threads, and adding `time.sleep` in main thread to preload the examples makes `preloaded_reader` example as fast as feed_dict one.

https://github.com/yaroslavvb/stuff/commit/226099cd883df5aa5b08483581d66960ada90e22

Note that simply increasing capacity/number of threads didn't help. I suspect Python just ran the main queue until it ran out of examples, and went back to slow behavior with blocking (Python3 specific thread behavior maybe?). 

Also, seems that there's no way to increase number of threads in `slice_input_producer`

**Comment by danielschonfeld** [comment]
@ebrevdo in my case, when producing timelines for with and without input ops, vs, feed_dict, it seems that the majority of the time is spent on a matmul operation in my `hidden_layer_1_1` which is the shared scope of my test.  (I obviously share my weights and biases in the hidden layers between my training data and my test data).

That operation, when used with input ops variables, takes 743ms!!!!!!

**Comment by danielschonfeld** [comment]
It appears that the MatMul operation takes a really long time since my test data size is longer than the training data batch size. (Training data is 100 example at each run while test is 3800 examples)

The question is why is sharing variables causing this size of a test data run to take so long as compared to utilizing place holders and not using variables scope sharing. 

Also important to note that overall with the smaller test batch size and variable sharing I'm still seeing about 100ms per run compared to 75ms per run with placeholders and a longer test size batch (3800 examples). 

But that's not as bad as my main problem which is 1000ms vs about 100ms

**Comment by danielschonfeld** [comment]
The situation is even worse when utilizing parse_single_example vs a QueueRunner with data preloaded and fed that way into the QueueRunner. 

With MatMul still taking a long time (for the reason stated in the above comment) and now in addition to that QueueDequeueMany taking a good 360ms. And that's with shuffle_batch, 4 threads, batch size of 100, capacity of 1000 and min dequeue of 200.

**Comment by danielschonfeld** [comment]
Just to add some info, my features tensor is about 32,000 features.  So the shape of the test set is (3800,32000) and the training data is usually of shape (100,32000).

Like i said, doing the above utilizing feed_dict takes about 80ms per `sess.run()` call and utilizing a QueueRunner with directly fed data items from numpy array (only for the training data, the test data is a const with `tf.convert_to_tensor()`) takes about 1000ms total with that matmul taking most of the time.  And finally utilizing a two `tfrecord` files, one for the training data and one for the test data, and feeding with `parse_single_example` for the training data and `parse_example` for the test data, `QueueDequeueMany` (<- of the test data, so the one using `parse_example`) and `Matmul` are the two highest time consumers bringing the `sess.run()` up to around 1800ms.

**Comment by yaroslavvb** [comment]
It sounds a bit suspicious that matmul itself would take different time depending on `feed_dict` vs `Queue`. Are you getting MatMul timing from the Timeline? Are you sure that matmul sizes are the same? Unlike with `QueueDequeue` many, IO time should not be included in the timeline timing, so if size is the same, the timing should also be the same

Here's [how](https://github.com/yaroslavvb/stuff/commit/aa886026ca3c48e27f776269548b40a2e2bb89ea) you can generate text file with timing information. In that file you can see duration of matmul as well as total output size for sanity checking

```
    node_stats {
      node_name: "hidden1/MatMul"
      all_start_micros: 1475724885315055
      op_start_rel_micros: 1
      op_end_rel_micros: 314
      all_end_rel_micros: 318
      memory {
        allocator_name: "cpu"
        total_bytes: 51200
      }

```

If the this timing information shows same total_bytes but 10x difference in timing, the way to get this triaged would be to provide a minimal self-contained reproducible test case

**Comment by danielschonfeld** [comment]
@yaroslavvb i'm certain about the size being the same.  But the difference between the two runs, except for being `feed_dict` vs `Queue` is that the matmul operation that takes long utilizing `Queue` is happening on the shared variable op node, an op that didn't exist (since it wasn't needed) on the `feed_dict` run.

I got the time from timeline, yes.  And i assume the time is what's called `wall time`. 

Finally, I'll go ahead and run it again, and save the raw RunMetadata this time.  I'll follow up with another comment shortly with the results of that.

**Comment by danielschonfeld** [comment]
@yaroslavvb so doing that revealed something I didn't realize before and that now i'm not sure how to proceed with.  

First things first, there's no problem with `Matmul`.  It produces the same amount of execution time for the same amount of bytes. I recorded only the first example from the two below, which in the feed_dict run didn't include the test data run.  I found that out by looking at the dim size and the output bytes as you suggested.

Basically I do something like the following:
`sess.run([optimizer, cost, summeryMerge])`
And later on I do something like `accuracy.eval()`

In the `Queue` based run, the value of accuracy is calculated at every run, even without me calling `accuracy.eval()` - I'm assuming because of the shared variables being updated since the graph has them built in.  That in turn is why my runs are considerably slower overall compared to `feed_dict`.  The operation takes the same amount of time, but is happening at every batch now instead of once per epoch.  I'm assuming that's also why the `Queue` that holds the test data gets depleted fast and takes a long while to continue producing sample data, causing `QueueDequeueMany` to appear slow.

My question then is how does one go about using `Queues` and feeding in test data but have it computed only once per epoch instead of at every batch run.  From reading `Sharing Variables`, I thought thats the way to go with training/test data.  So I have a structure similar to the following code:

```
def neural_network_model(x):
    with tf.variable_scope("hidden_layer_1"):
        relu1 = hl_relu(x, [max_words_len, n_nodes_hl1], [n_nodes_hl1]) #hl_relu utilizes get_variable for weights and biases
```
And later on i'd do something like this:
```
def train_neural_network(x_batch, y_batch, test_x, test_y):
    with tf.device("/cpu:0"):
        with tf.variable_scope("dnn") as scope:
            logits = neural_network_model(x_batch)
            scope.reuse_variables()
            test_logits = neural_network_model(test_x)
```

I'd use `logits` to compute the cost with `softmax_cross_entropy_with_logits`, as follows:
```
cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(logits, y_batch), name="cost")
```

and `test_logits` to compute the accuracy in a three step operation looking something like this:
```
prediction = tf.nn.softmax(test_logits)
correct = tf.equal(tf.argmax(prediction, 1), tf.argmax(test_y, 1), name="correct")
accuracy = tf.reduce_mean(tf.cast(correct, tf.float32), name="accuracy")
```

What am I doing wrong?

**Comment by danielschonfeld** [comment]
Maybe i should just simplify my question.  How do I train and test in the same process, using shared variables with 2 tfrecord files, one designated as train data and one designated as test data.

As mentioned briefly in on the last line of [https://www.tensorflow.org/versions/r0.11/how_tos/reading_data/index.html#multiple-input-pipelines](url)

**Comment by ebrevdo** [comment]
I'm confused why matmul timing would change unless the shapes changed, or
perhaps if it for some reason started running on a CPU device where before
it was running on a GPU device.  Please provide the runmetadata.

**Comment by danielschonfeld** [comment]
@ebrevdo the shape changed... it was my mistake.  My timing comparison was incorrect.  When compared with the same shape the timing was identical.

So now I'm just left wondering, what I asked above.  How can I do both train and eval in the same process using input ops?

**Comment by ebrevdo** [comment]
It's atypical to do this.  Folks usually build two graphs: one for
training, and a separate one for eval.  (you can use a common function to
build the "common" subgraph).  You then run the train and eval graphs
either in separate threads or in different processes.  At the very least,
you use different sessions for training and eval.

**Comment by danielschonfeld** [comment]
@ebrevdo I see... so i should adopt more of like the cifar-10 example model where one process trains and one monitors for checkpoint files and loads and tries the eval stuff against?

**Comment by ebrevdo** [comment]
Exactly. Then you keep the best checkpoint.

**Comment by danielschonfeld** [comment]
now if only i can figure out why my raw usage of tensorflow only achieves 60% accuracy compared to using the SciKit module of TF `DNNClassifier` which archives like 99% pretty fast.

Perhaps the usage of embedded lookup? We can probably close this issue for now, unless someone else wants to keep it alive?

**Comment by gunan** [comment]
Going through the issues mentioning r0.11.
Looks like the initial issue is resolved here. Closing.
